plugins/BEdita/Core/src/Model/Behavior/CustomPropertiesBehavior.php

```diff
@@ -0,0 +1,160 @@
+            $props = TableRegistry::get('Properties')->find()
+                ->where([
+                    'object_type_id' => $objectType->id,
+                    'enabled' => 1,
+                ])
+                ->all();
+        } catch (\Exception $ex) {
+            $props = false;
+        }
+
+        if (!$props) {
+            return $this->available = [];
+        }
+
+        $this->available = $props->combine('name', function ($entity) {
+            return $entity;
+        })->toArray();
+
+        return $this->available;
+    }
+
+    /**
+     * Return the default values of available properties
+     *
+     * @return array
+     */
+    public function getDefaultValues()
+    {
+        return array_fill_keys(array_keys($this->getAvailable()), null);
+    }
+
+    /**
+     * Set custom properties keys as main properties
+     *
+     * @param \Cake\Event\Event $event The beforeFind event that was fired.
+     * @param \Cake\ORM\Query $query Query
+     * @return void
+     */
+    public function beforeFind(Event $event, Query $query)
+    {
+        $query->formatResults(function ($results) {
+            return $results->map(function ($row) {
+                return $this->promoteProperties($row);
+            });
+        });
+    }
+
+    /**
+     * Promote the properties in configuration `field` to first citizen property.
+     * Missing properties in `$entity` but available will be filled with default values.
+     *
+     * @param \Cake\Datasource\EntityInterface|array $entity The entity or the array to work on
+     * @return \Cake\Datasource\EntityInterface|array
+     */
+    protected function promoteProperties($entity)
+    {
+        if ((!is_array($entity) && !($entity instanceof EntityInterface)) || !$this->isFieldSet($entity)) {
+            return $entity;
+        }
+
+        $field = $this->getConfig('field');
+        if (empty($entity[$field]) || !is_array($entity[$field])) {
+            $entity[$field] = [];
+        }
+        $entity[$field] = $entity[$field] + $this->getDefaultValues();
+
+        if (empty($entity[$field])) {
+            return $entity;
+        }
+
+        $customProps = $entity[$field];
+        unset($entity[$field]);
+
+        if (is_array($entity)) {
+            return array_merge($entity, $customProps);
+        }
+
+        $entity->set($customProps, ['guard' => false])->clean();
+
+        return $entity;
+    }
+
+    /**
+     * Check if configured field containing custom properties is set in `$entity`.
+     * For "set" is intended that it is present in `$entity` with any value.
+     *
+     * @param \Cake\Datasource\EntityInterface|array $entity The entity or the array to check
+     * @return bool
+     */
+    protected function isFieldSet($entity)
+    {
+        $field = $this->getConfig('field');
+
+        $allProperties = $entity;
+        if ($entity instanceof EntityInterface) {
+            $allProperties = clone($entity);
```

Comment:
`clone` is a language construct just like `echo`, `require` and `exit`. Parentheses around the "argument" are not necessary and usually omitted.

Reply:
done